File: tests/test_lat_long_system.cpp
#include <gtest/gtest.h>

#include "quad_key/geo_coordinate2d.h"
#include "quad_key/quad_key.h"

#include <iostream>

namespace quad_keys
{
  class test_lat_long_system : public ::testing::Test
  {
  protected:
    virtual void SetUp() {}
    virtual void TearDown() {}
  };

  TEST(LATLONGSYSTEM, is_valid)
  {
    // Same as bing_system
  }

  TEST(LATLONGSYSTEM, is_root_key)
  {
    // Same as bing_system
  }

  TEST(LATLONGSYSTEM, get_parent)
  {
    // Same as bing_system
  }

  TEST(LATLONGSYSTEM, is_my_ancestor)
  {
    // Same as bing_system
  }

  TEST(LATLONGSYSTEM, get_ancestor)
  {
    // Same as bing_system
  }

  TEST(LATLONGSYSTEM, get_child)
  {
    // Same as bing_system
  }

  TEST(LATLONGSYSTEM, get_children)
  {
    // Same as bing_system
  }

  TEST(LATLONGSYSTEM, get_descendants)
  {
    // Same as bing_system
  }

  TEST(LATLONGSYSTEM, get_siblings)
  {
    // Same as bing_system
  }

  TEST(LATLONGSYSTEM, get_neighbors)
  {
    // Same as bing_system
  }

  TEST(LATLONGSYSTEM, from_string_to_string)
  {
    // Same as bing_system
  }

  TEST(LATLONGSYSTEM, get_geocoordinate_bounds2d)
  {
    quad_key key;
    geo_coordinate_bounding_box2d bbox;

    { // Depth 0
      key = quad_key(type::lat_long, 0, 0, 0);
      key.get_geo_coordinate_bounds2d(bbox);
      EXPECT_EQ(geo_coordinate2d(longitude(-180.0), latitude(-90.0)),
        bbox.lower_left());
      EXPECT_EQ(geo_coordinate2d(longitude(+180.0), latitude(+90.0)),
        bbox.upper_right());
    }

    { // Depth 1
      key = quad_key(type::lat_long, 0, 0, 1);
      key.get_geo_coordinate_bounds2d(bbox);
      EXPECT_TRUE(bbox.lower_left().get_longitude().near(
        longitude(-180.0), std::numeric_limits<float>::epsilon()));
      EXPECT_TRUE(bbox.lower_left().get_latitude().near(
        latitude(-90.0), std::numeric_limits<float>::epsilon()));
      EXPECT_TRUE(bbox.upper_right().get_longitude().near(
        longitude(+0.0), std::numeric_limits<float>::epsilon()));
      EXPECT_TRUE(bbox.upper_right().get_latitude().near(
        latitude(+0.0), std::numeric_limits<float>::epsilon()));

      key = quad_key(type::lat_long, 0, 1, 1);
      key.get_geo_coordinate_bounds2d(bbox);
      EXPECT_TRUE(bbox.lower_left().get_longitude().near(
        longitude(+0.0), std::numeric_limits<float>::epsilon()));
      EXPECT_TRUE(bbox.lower_left().get_latitude().near(
        latitude(-90.0), std::numeric_limits<float>::epsilon()));
      EXPECT_TRUE(bbox.upper_right().get_longitude().near(
        longitude(+180.0), std::numeric_limits<float>::epsilon()));
      EXPECT_TRUE(bbox.upper_right().get_latitude().near(
        latitude(+0.0), std::numeric_limits<float>::epsilon()));

      key = quad_key(type::lat_long, 1, 0, 1);
      key.get_geo_coordinate_bounds2d(bbox);
      EXPECT_TRUE(bbox.lower_left().get_longitude().near(
        longitude(-180.0), std::numeric_limits<float>::epsilon()));
      EXPECT_TRUE(bbox.lower_left().get_latitude().near(
        latitude(+0.0), std::numeric_limits<float>::epsilon()));
      EXPECT_TRUE(bbox.upper_right().get_longitude().near(
        longitude(+0.0), std::numeric_limits<float>::epsilon()));
      EXPECT_TRUE(bbox.upper_right().get_latitude().near(
        latitude(+90.0), std::numeric_limits<float>::epsilon()));

      key = quad_key(type::lat_long, 1, 1, 1);
      key.get_geo_coordinate_bounds2d(bbox);
      EXPECT_TRUE(bbox.lower_left().get_longitude().near(
        longitude(+0.0), std::numeric_limits<float>::epsilon()));
      EXPECT_TRUE(bbox.lower_left().get_latitude().near(
        latitude(+0.0), std::numeric_limits<float>::epsilon()));
      EXPECT_TRUE(bbox.upper_right().get_longitude().near(
        longitude(+180.0), std::numeric_limits<float>::epsilon()));
      EXPECT_TRUE(bbox.upper_right().get_latitude().near(
        latitude(+90.0), std::numeric_limits<float>::epsilon()));
    }
  }

  TEST(LATLONGSYSTEM, get_key_from_longitude_latitude_at_depth_to_bounds)
  {
     quad_key key;

      { // Depth 0
        key = quad_key::get_key_from_longitude_latitude_at_depth(
            type::lat_long, geo_coordinate2d(longitude(-180.0),
            latitude(-90)), 0u);
        EXPECT_EQ(quad_key(type::lat_long, 0, 0, 0), key);
        EXPECT_EQ("", key.to_internal_string());
      }

      // Inclusion here consists of lower left
      //    X------X------X
      //    |   2  |   3  |
      //    |      |      |
      //    X------X------X
      //    |   0  |   1  |
      //    |      |      |
      // ll X------X------X
      { // Depth 1
        key = quad_key::get_key_from_longitude_latitude_at_depth( // Q0 ll
          type::lat_long, geo_coordinate2d(longitude(-180.0),
          latitude(-90.0)), 1u);
        EXPECT_EQ(quad_key(type::lat_long, 0, 0, 1), key);
        EXPECT_EQ("2", key.to_internal_string());
        key = quad_key::get_key_from_longitude_latitude_at_depth( // Q0 lr
          type::lat_long, geo_coordinate2d(longitude(-0.1),
          latitude(-90.0)), 1u);
        EXPECT_EQ(quad_key(type::lat_long, 0, 0, 1), key);
        EXPECT_EQ("2", key.to_internal_string());
        key = quad_key::get_key_from_longitude_latitude_at_depth( // Q0 ur
          type::lat_long, geo_coordinate2d(longitude(-0.1),
          latitude(-0.1)), 1u);
        EXPECT_EQ(quad_key(type::lat_long, 0, 0, 1), key);
        EXPECT_EQ("2", key.to_internal_string());
        key = quad_key::get_key_from_longitude_latitude_at_depth( // Q0 ul
          type::lat_long, geo_coordinate2d(longitude(-180.0),
          latitude(-0.1)), 1u);
        EXPECT_EQ(quad_key(type::lat_long, 0, 0, 1), key);
        EXPECT_EQ("2", key.to_internal_string());

        key = quad_key::get_key_from_longitude_latitude_at_depth( // Q1 ll
          type::lat_long, geo_coordinate2d(longitude(+0.0),
          latitude(-90.0)), 1u);
        EXPECT_EQ(quad_key(type::lat_long, 0, 1, 1), key);
        EXPECT_EQ("3", key.to_internal_string());
        key = quad_key::get_key_from_longitude_latitude_at_depth( // Q1 lr
          type::lat_long, geo_coordinate2d(longitude(+180.0),
          latitude(-90.0)), 1u);
        EXPECT_EQ(quad_key(type::lat_long, 0, 1, 1), key);
        EXPECT_EQ("3", key.to_internal_string());
        key = quad_key::get_key_from_longitude_latitude_at_depth( // Q1 ur
          type::lat_long, geo_coordinate2d(longitude(+180.0),
          latitude(-0.1)), 1u);
        EXPECT_EQ(quad_key(type::lat_long, 0, 1, 1), key);
        EXPECT_EQ("3", key.to_internal_string());
        key = quad_key::get_key_from_longitude_latitude_at_depth( // Q1 ul
          type::lat_long, geo_coordinate2d(longitude(+0.0),
          latitude(-0.1)), 1u);
        EXPECT_EQ(quad_key(type::lat_long, 0, 1, 1), key);
        EXPECT_EQ("3", key.to_internal_string());

        key = quad_key::get_key_from_longitude_latitude_at_depth( // Q2 ll
          type::lat_long, geo_coordinate2d(longitude(-180.0),
          latitude(+0.0)), 1u);
        EXPECT_EQ(quad_key(type::lat_long, 1, 0, 1), key);
        EXPECT_EQ("0", key.to_internal_string());
        key = quad_key::get_key_from_longitude_latitude_at_depth( // Q2 lr
          type::lat_long, geo_coordinate2d(longitude(-0.1),
          latitude(+0.0)), 1u);
        EXPECT_EQ(quad_key(type::lat_long, 1, 0, 1), key);
        EXPECT_EQ("0", key.to_internal_string());
        key = quad_key::get_key_from_longitude_latitude_at_depth( // Q2 ur
          type::lat_long, geo_coordinate2d(longitude(-180.0),
          latitude(+90.0)), 1u);
        EXPECT_EQ(quad_key(type::lat_long, 1, 0, 1), key);
        EXPECT_EQ("0", key.to_internal_string());
        key = quad_key::get_key_from_longitude_latitude_at_depth( // Q2 ul
          type::lat_long, geo_coordinate2d(longitude(-0.1),
          latitude(+90.0)), 1u);
        EXPECT_EQ(quad_key(type::lat_long, 1, 0, 1), key);
        EXPECT_EQ("0", key.to_internal_string());

        key = quad_key::get_key_from_longitude_latitude_at_depth( // Q3 ll
          type::lat_long, geo_coordinate2d(longitude(+0.0),
          latitude(+0.0)), 1u);
        EXPECT_EQ(quad_key(type::lat_long, 1, 1, 1), key);
        EXPECT_EQ("1", key.to_internal_string());
        key = quad_key::get_key_from_longitude_latitude_at_depth( // Q3 lr
          type::lat_long, geo_coordinate2d(longitude(+180.0),
          latitude(+0.0)), 1u);
        EXPECT_EQ(quad_key(type::lat_long, 1, 1, 1), key);
        EXPECT_EQ("1", key.to_internal_string());
        key = quad_key::get_key_from_longitude_latitude_at_depth( // Q3 ur
          type::lat_long, geo_coordinate2d(longitude(+0.0),
          latitude(+90.0)), 1u);
        EXPECT_EQ(quad_key(type::lat_long, 1, 1, 1), key);
        EXPECT_EQ("1", key.to_internal_string());
        key = quad_key::get_key_from_longitude_latitude_at_depth( // Q3 ul
          type::lat_long, geo_coordinate2d(longitude(+180.0),
          latitude(+90.0)), 1u);
        EXPECT_EQ(quad_key(type::lat_long, 1, 1, 1), key);
        EXPECT_EQ("1", key.to_internal_string());
      }
  }

  TEST(LATLONGSYSTEM, get_keys_from_longitude_latitude_bounding_box)
  {
    {
      std::vector<quad_key> out_keys;
      out_keys = quad_key::get_keys_from_longitude_latitude_bounding_box(
        type::lat_long,
        geo_coordinate_bounding_box2d(
          geo_coordinate2d(longitude(-180.0), latitude(+90.0)),
          geo_coordinate2d(longitude(+180.0), latitude(-90.0))), 0);
      EXPECT_EQ(1u, out_keys.size());
      EXPECT_NE(out_keys.end(), std::find(out_keys.begin(), out_keys.end(),
        quad_key(type::lat_long, 0, 0, 0)));
    }

    {
      std::vector<quad_key> out_keys;
      out_keys = quad_key::get_keys_from_longitude_latitude_bounding_box(
        type::lat_long,
        geo_coordinate_bounding_box2d(
          geo_coordinate2d(longitude(-180.0), latitude(+90.0)),
          geo_coordinate2d(longitude(+180.0), latitude(-90.0))), 1);
      EXPECT_EQ(4u, out_keys.size());
      EXPECT_NE(out_keys.end(), std::find(out_keys.begin(), out_keys.end(),
        quad_key(type::lat_long, 0, 0, 1)));
      EXPECT_NE(out_keys.end(), std::find(out_keys.begin(), out_keys.end(),
        quad_key(type::lat_long, 0, 1, 1)));
      EXPECT_NE(out_keys.end(), std::find(out_keys.begin(), out_keys.end(),
        quad_key(type::lat_long, 1, 0, 1)));
      EXPECT_NE(out_keys.end(), std::find(out_keys.begin(), out_keys.end(),
        quad_key(type::lat_long, 1, 1, 1)));
    }

    {
      std::vector<quad_key> out_keys;
      out_keys = quad_key::get_keys_from_longitude_latitude_bounding_box(
        type::lat_long,
        geo_coordinate_bounding_box2d(
          geo_coordinate2d(longitude(-180.0), latitude(-0.1)),
          geo_coordinate2d(longitude(-0.1), latitude(+90.0))), 1);
      EXPECT_EQ(2u, out_keys.size());
      EXPECT_NE(out_keys.end(), std::find(out_keys.begin(), out_keys.end(),
        quad_key(type::lat_long, 0, 0, 1)));
      EXPECT_NE(out_keys.end(), std::find(out_keys.begin(), out_keys.end(),
        quad_key(type::lat_long, 1, 0, 1)));
    }

    {
      std::vector<quad_key> out_keys;
      out_keys = quad_key::get_keys_from_longitude_latitude_bounding_box(
        type::lat_long,
        geo_coordinate_bounding_box2d(
          geo_coordinate2d(longitude(+0.0), latitude(-0.1)),
          geo_coordinate2d(longitude(+180.0), latitude(+90.0))), 1);
      EXPECT_EQ(2u, out_keys.size());
      EXPECT_NE(out_keys.end(), std::find(out_keys.begin(), out_keys.end(),
        quad_key(type::lat_long, 0, 1, 1)));
      EXPECT_NE(out_keys.end(), std::find(out_keys.begin(), out_keys.end(),
        quad_key(type::lat_long, 1, 1, 1)));
    }

    {
      std::vector<quad_key> out_keys;
      out_keys = quad_key::get_keys_from_longitude_latitude_bounding_box(
        type::lat_long,
        geo_coordinate_bounding_box2d(
          geo_coordinate2d(longitude(-0.1), latitude(-0.1)),
          geo_coordinate2d(longitude(+0.1), latitude(+0.1))), 1);
      EXPECT_EQ(4u, out_keys.size());
      EXPECT_NE(out_keys.end(), std::find(out_keys.begin(), out_keys.end(),
          quad_key(type::lat_long, 0, 0, 1)));
      EXPECT_NE(out_keys.end(), std::find(out_keys.begin(), out_keys.end(),
          quad_key(type::lat_long, 0, 1, 1)));
      EXPECT_NE(out_keys.end(), std::find(out_keys.begin(), out_keys.end(),
          quad_key(type::lat_long, 1, 0, 1)));
      EXPECT_NE(out_keys.end(), std::find(out_keys.begin(), out_keys.end(),
          quad_key(type::lat_long, 1, 1, 1)));
    }
  }

  TEST(LATLONGSYSTEM, test_known_sample_of_locations)
  {
    quad_key key;

    {
      key = quad_key::get_key_from_longitude_latitude_at_depth(
        type::lat_long, geo_coordinate2d(longitude(-122.4183),
        latitude(37.7750)), 1u);
      EXPECT_EQ("0", key.to_internal_string());
      key = quad_key::get_key_from_longitude_latitude_at_depth(
        type::lat_long, geo_coordinate2d(longitude(-122.4183),
        latitude(37.7750)), 2u);
      EXPECT_EQ("02", key.to_internal_string());
      key = quad_key::get_key_from_longitude_latitude_at_depth(
        type::lat_long, geo_coordinate2d(longitude(-122.4183),
        latitude(37.7750)), 3u);
      EXPECT_EQ("021", key.to_internal_string());
      key = quad_key::get_key_from_longitude_latitude_at_depth(
        type::lat_long, geo_coordinate2d(longitude(-122.4183),
        latitude(37.7750)), 4u);
      EXPECT_EQ("0210", key.to_internal_string());
    }

    {
      key = quad_key::get_key_from_longitude_latitude_at_depth(
        type::lat_long, geo_coordinate2d(longitude(+134.9),
        latitude(-67.6)), 3u);
      EXPECT_EQ(quad_key(type::lat_long, 0, 6, 3), key);
      EXPECT_EQ("332", key.to_internal_string());

      key = quad_key::get_key_from_longitude_latitude_at_depth(
        type::lat_long, geo_coordinate2d(longitude(+135.0),
        latitude(-67.5)), 3u);
      EXPECT_EQ(quad_key(type::lat_long, 1, 7, 3), key);
      EXPECT_EQ("331", key.to_internal_string());
    }
  }
}

File: bindings/python/tests/test_lat_long_system.py
"""
Unit Tests for LatLong System
"""

from __future__ import print_function
import unittest
from quadkey import QuadKey
from quadkey import Type_LatLong
from quadkey import GeoCoordinate2d
from quadkey import GeoCoordinateBoundingBox2d
from quadkey import Longitude
from quadkey import Latitude


class TestLatLongSystem(unittest.TestCase):
    """
    Class for running unit tests.
    """

    def test_is_valid(self):
        """
        See TestQuadKey.test_is_valid
        """

    def test_is_root_key(self):
        """
        See TestQuadKey.test_is_root_key
        """

    def test_get_parent(self):
        """
        Test to get parent of a generically formatted quad_key.
        """
        # Same as BingSystem

    def test_is_my_ancestor(self):
        """
        Checks to see if specified quadkey up in the hierarchy is an "ancestor"
        of the given quad_key.
        """
        # Same as BingSystem

    def test_get_ancestor(self):
        """
        Gets the ancestor up the tree according to a delta provided.
        """
        # Same as BingSystem

    def test_get_child(self):
        """
        Get a single child NE, SE, NW, SW
        """
        # Same as BingSystem

    def test_get_children(self):
        """
        Get all the childern of a quad_key.
        """
        # Same as BingSystem

    def test_get_descendants(self):
        """
        Get all the defendants at some delta in the tree.
        """
        # Same as BingSystem

    def test_get_siblings(self):
        """
        Test to get siblings of a given quadkey
        """
        # Same as BingSystem

    def test_get_neighbors(self):
        """
        Test to get the neighbors of a given quadquad_key.
        """
        # Same as BingSystem

    def test_to_string_from_string(self):
        """
        Test to convert an internally formatted string and back.
        """
        # Same as BingSystem

    def test_get_geo_coordinate_bounds_2d(self):
        """
        Test to get geographic bounds of quadquad_key.
        """
        bbox = GeoCoordinateBoundingBox2d()

        quad_key = QuadKey(Type_LatLong, 0, 0, 0)
        quad_key.getGeoCoordinateBounds2d(bbox)
        self.assertTrue(GeoCoordinate2d(Longitude(-180.0), Latitude(-90.0)) ==
                        bbox.lowerLeft())
        self.assertTrue(GeoCoordinate2d(Longitude(+180.0), Latitude(+90.0)) ==
                        bbox.upperRight())

        quad_key = QuadKey(Type_LatLong, 0, 0, 1)
        quad_key.getGeoCoordinateBounds2d(bbox)
        self.assertTrue(bbox.lowerLeft().longitude().near(Longitude(-180.0),
                                                          1e-6))
        self.assertTrue(bbox.lowerLeft().latitude().near(Latitude(-90.0),
                                                         1e-6))
        self.assertTrue(bbox.upperRight().longitude().near(Longitude(+0.0),
                                                           1e-6))
        self.assertTrue(bbox.upperRight().latitude().near(Latitude(+0.0),
                                                          1e-6))

        quad_key = QuadKey(Type_LatLong, 0, 1, 1)
        quad_key.getGeoCoordinateBounds2d(bbox)
        self.assertTrue(bbox.lowerLeft().longitude().near(Longitude(+0.0),
                                                          1e-6))
        self.assertTrue(bbox.lowerLeft().latitude().near(Latitude(-90.0),
                                                         1e-6))
        self.assertTrue(bbox.upperRight().longitude().near(Longitude(+180.0),
                                                           1e-6))
        self.assertTrue(bbox.upperRight().latitude().near(Latitude(+0.0),
                                                          1e-6))

        quad_key = QuadKey(Type_LatLong, 1, 0, 1)
        quad_key.getGeoCoordinateBounds2d(bbox)
        self.assertTrue(bbox.lowerLeft().longitude().near(Longitude(-180.0),
                                                          1e-6))
        self.assertTrue(bbox.lowerLeft().latitude().near(Latitude(+0.0),
                                                         1e-6))
        self.assertTrue(bbox.upperRight().longitude().near(Longitude(+0.0),
                                                           1e-6))
        self.assertTrue(bbox.upperRight().latitude().near(Latitude(+90.0),
                                                          1e-6))

        quad_key = QuadKey(Type_LatLong, 1, 1, 1)
        quad_key.getGeoCoordinateBounds2d(bbox)
        self.assertTrue(bbox.lowerLeft().longitude().near(Longitude(+0.0),
                                                          1e-6))
        self.assertTrue(bbox.lowerLeft().latitude().near(Latitude(+0.0),
                                                         1e-6))
        self.assertTrue(bbox.upperRight().longitude().near(Longitude(+180.0),
                                                           1e-6))
        self.assertTrue(bbox.upperRight().latitude().near(Latitude(+90.0),
                                                          1e-6))

    def test_get_key_from_longitude_latitude_at_depth_to_bounds(self):
        """
        Test to get a quadkey from longitude and latitude and then get bounds.
        """

        quad_key = QuadKey.getKeyFromLongitudeLatitudeAtDepth(
            Type_LatLong, GeoCoordinate2d(Longitude(-180.0),
                                          Latitude(-90)), 0)
        self.assertTrue(QuadKey(Type_LatLong, 0, 0, 0) == quad_key)
        self.assertEquals("", quad_key.toInternalString())

        # Inclusion here consists of lower left
        #    X------X------X
        #    |   2  |   3  |
        #    |      |      |
        #    X------X------X
        #    |   0  |   1  |
        #    |      |      |
        # ll X------X------X
        quad_key = QuadKey.getKeyFromLongitudeLatitudeAtDepth(
            Type_LatLong, GeoCoordinate2d(Longitude(-180.0),
                                          Latitude(-90.0)), 1)
        self.assertTrue(QuadKey(Type_LatLong, 0, 0, 1) == quad_key)
        self.assertEquals("2", quad_key.toInternalString())
        quad_key = QuadKey.getKeyFromLongitudeLatitudeAtDepth(
            Type_LatLong, GeoCoordinate2d(Longitude(-0.1),
                                          Latitude(-90.0)), 1)
        self.assertTrue(QuadKey(Type_LatLong, 0, 0, 1) == quad_key)
        self.assertEquals("2", quad_key.toInternalString())
        quad_key = QuadKey.getKeyFromLongitudeLatitudeAtDepth(
            Type_LatLong, GeoCoordinate2d(Longitude(-0.1),
                                          Latitude(-0.1)), 1)
        self.assertTrue(QuadKey(Type_LatLong, 0, 0, 1) == quad_key)
        self.assertEquals("2", quad_key.toInternalString())
        quad_key = QuadKey.getKeyFromLongitudeLatitudeAtDepth(
            Type_LatLong, GeoCoordinate2d(Longitude(-180.0),
                                          Latitude(-0.1)), 1)
        self.assertTrue(QuadKey(Type_LatLong, 0, 0, 1) == quad_key)
        self.assertEquals("2", quad_key.toInternalString())

        quad_key = QuadKey.getKeyFromLongitudeLatitudeAtDepth(
            Type_LatLong, GeoCoordinate2d(Longitude(+0.0),
                                          Latitude(-90.0)), 1)
        self.assertTrue(QuadKey(Type_LatLong, 0, 1, 1) == quad_key)
        self.assertEquals("3", quad_key.toInternalString())
        quad_key = QuadKey.getKeyFromLongitudeLatitudeAtDepth(
            Type_LatLong, GeoCoordinate2d(Longitude(+180.0),
                                          Latitude(-90.0)), 1)
        self.assertTrue(QuadKey(Type_LatLong, 0, 1, 1) == quad_key)
        self.assertEquals("3", quad_key.toInternalString())
        quad_key = QuadKey.getKeyFromLongitudeLatitudeAtDepth(
            Type_LatLong, GeoCoordinate2d(Longitude(+180.0),
                                          Latitude(-0.1)), 1)
        self.assertTrue(QuadKey(Type_LatLong, 0, 1, 1) == quad_key)
        self.assertEquals("3", quad_key.toInternalString())
        quad_key = QuadKey.getKeyFromLongitudeLatitudeAtDepth(
            Type_LatLong, GeoCoordinate2d(Longitude(+0.0),
                                          Latitude(-0.1)), 1)
        self.assertTrue(QuadKey(Type_LatLong, 0, 1, 1) == quad_key)
        self.assertEquals("3", quad_key.toInternalString())

        quad_key = QuadKey.getKeyFromLongitudeLatitudeAtDepth(
            Type_LatLong, GeoCoordinate2d(Longitude(-180.0),
                                          Latitude(+0.0)), 1)
        self.assertTrue(QuadKey(Type_LatLong, 1, 0, 1) == quad_key)
        self.assertEquals("0", quad_key.toInternalString())
        quad_key = QuadKey.getKeyFromLongitudeLatitudeAtDepth(
            Type_LatLong, GeoCoordinate2d(Longitude(-0.1),
                                          Latitude(+0.0)), 1)
        self.assertTrue(QuadKey(Type_LatLong, 1, 0, 1) == quad_key)
        self.assertEquals("0", quad_key.toInternalString())
        quad_key = QuadKey.getKeyFromLongitudeLatitudeAtDepth(
            Type_LatLong, GeoCoordinate2d(Longitude(-180.0),
                                          Latitude(+90.0)), 1)
        self.assertTrue(QuadKey(Type_LatLong, 1, 0, 1) == quad_key)
        self.assertEquals("0", quad_key.toInternalString())
        quad_key = QuadKey.getKeyFromLongitudeLatitudeAtDepth(
            Type_LatLong, GeoCoordinate2d(Longitude(-0.1),
                                          Latitude(+90.0)), 1)
        self.assertTrue(QuadKey(Type_LatLong, 1, 0, 1) == quad_key)
        self.assertEquals("0", quad_key.toInternalString())

        quad_key = QuadKey.getKeyFromLongitudeLatitudeAtDepth(
            Type_LatLong, GeoCoordinate2d(Longitude(+0.0),
                                          Latitude(+0.0)), 1)
        self.assertTrue(QuadKey(Type_LatLong, 1, 1, 1) == quad_key)
        self.assertEquals("1", quad_key.toInternalString())
        quad_key = QuadKey.getKeyFromLongitudeLatitudeAtDepth(
            Type_LatLong, GeoCoordinate2d(Longitude(+180.0),
                                          Latitude(+0.0)), 1)
        self.assertTrue(QuadKey(Type_LatLong, 1, 1, 1) == quad_key)
        self.assertEquals("1", quad_key.toInternalString())
        quad_key = QuadKey.getKeyFromLongitudeLatitudeAtDepth(
            Type_LatLong, GeoCoordinate2d(Longitude(+0.0),
                                          Latitude(+90.0)), 1)
        self.assertTrue(QuadKey(Type_LatLong, 1, 1, 1) == quad_key)
        self.assertEquals("1", quad_key.toInternalString())
        quad_key = QuadKey.getKeyFromLongitudeLatitudeAtDepth(
            Type_LatLong, GeoCoordinate2d(Longitude(+180.0),
                                          Latitude(+90.0)), 1)
        self.assertTrue(QuadKey(Type_LatLong, 1, 1, 1) == quad_key)
        self.assertEquals("1", quad_key.toInternalString())

    def test_get_keys_from_longitude_latitude_bounding_box(self):
        """
        Test to get the bounding box of a quadquad_key.
        """
        outkeys = QuadKey.getKeysFromLongitudeLatitudeBoundingBox(
            Type_LatLong,
            GeoCoordinateBoundingBox2d(
                GeoCoordinate2d(Longitude(-180.0), Latitude(+90.0)),
                GeoCoordinate2d(Longitude(+180.0), Latitude(-90.0))),
            0)
        self.assertEquals(1, outkeys.size())
        self.assertTrue(QuadKey(Type_LatLong, 0, 0, 0) in outkeys)

        outkeys = QuadKey.getKeysFromLongitudeLatitudeBoundingBox(
            Type_LatLong,
            GeoCoordinateBoundingBox2d(
                GeoCoordinate2d(Longitude(-180.0), Latitude(+90.0)),
                GeoCoordinate2d(Longitude(+180.0), Latitude(-90.0))),
            1)
        self.assertEquals(4, outkeys.size())
        self.assertTrue(QuadKey(Type_LatLong, 0, 0, 1) in outkeys)
        self.assertTrue(QuadKey(Type_LatLong, 0, 1, 1) in outkeys)
        self.assertTrue(QuadKey(Type_LatLong, 1, 0, 1) in outkeys)
        self.assertTrue(QuadKey(Type_LatLong, 1, 1, 1) in outkeys)

        outkeys = QuadKey.getKeysFromLongitudeLatitudeBoundingBox(
            Type_LatLong,
            GeoCoordinateBoundingBox2d(
                GeoCoordinate2d(Longitude(-180.0), Latitude(-0.1)),
                GeoCoordinate2d(Longitude(-0.1), Latitude(+90.0))),
            1)
        self.assertEquals(2, outkeys.size())
        self.assertTrue(QuadKey(Type_LatLong, 0, 0, 1) in outkeys)
        self.assertTrue(QuadKey(Type_LatLong, 1, 0, 1) in outkeys)

        outkeys = QuadKey.getKeysFromLongitudeLatitudeBoundingBox(
            Type_LatLong,
            GeoCoordinateBoundingBox2d(
                GeoCoordinate2d(Longitude(+0.0), Latitude(-0.1)),
                GeoCoordinate2d(Longitude(+180.0), Latitude(+90.0))),
            1)
        self.assertEquals(2, outkeys.size())
        self.assertTrue(QuadKey(Type_LatLong, 0, 1, 1) in outkeys)
        self.assertTrue(QuadKey(Type_LatLong, 1, 1, 1) in outkeys)

        outkeys = QuadKey.getKeysFromLongitudeLatitudeBoundingBox(
            Type_LatLong,
            GeoCoordinateBoundingBox2d(
                GeoCoordinate2d(Longitude(-0.1), Latitude(-0.1)),
                GeoCoordinate2d(Longitude(+0.1), Latitude(+0.1))),
            1)
        self.assertEquals(4, outkeys.size())
        self.assertTrue(QuadKey(Type_LatLong, 0, 0, 1) in outkeys)
        self.assertTrue(QuadKey(Type_LatLong, 0, 1, 1) in outkeys)
        self.assertTrue(QuadKey(Type_LatLong, 1, 0, 1) in outkeys)
        self.assertTrue(QuadKey(Type_LatLong, 1, 1, 1) in outkeys)

    def test_selected_cities(self):
        """
        Test a given set of known places make sure they work.
        """
        quad_key = QuadKey.getKeyFromLongitudeLatitudeAtDepth(
            Type_LatLong, GeoCoordinate2d(Longitude(-122.4183),
                                          Latitude(37.7750)), 1)
        self.assertEquals("0", quad_key.toInternalString())
        quad_key = QuadKey.getKeyFromLongitudeLatitudeAtDepth(
            Type_LatLong, GeoCoordinate2d(Longitude(-122.4183),
                                          Latitude(37.7750)), 2)
        self.assertEquals("02", quad_key.toInternalString())
        quad_key = QuadKey.getKeyFromLongitudeLatitudeAtDepth(
            Type_LatLong, GeoCoordinate2d(Longitude(-122.4183),
                                          Latitude(37.7750)), 3)
        self.assertEquals("021", quad_key.toInternalString())
        quad_key = QuadKey.getKeyFromLongitudeLatitudeAtDepth(
            Type_LatLong, GeoCoordinate2d(Longitude(-122.4183),
                                          Latitude(37.7750)), 4)
        self.assertEquals("0210", quad_key.toInternalString())

        quad_key = QuadKey.getKeyFromLongitudeLatitudeAtDepth(
            Type_LatLong, GeoCoordinate2d(Longitude(+134.9),
                                          Latitude(-67.6)), 3)
        self.assertTrue(QuadKey(Type_LatLong, 0, 6, 3) == quad_key)
        self.assertEquals("332", quad_key.toInternalString())

        quad_key = QuadKey.getKeyFromLongitudeLatitudeAtDepth(
            Type_LatLong, GeoCoordinate2d(Longitude(+135.0),
                                          Latitude(-67.5)), 3)
        self.assertTrue(QuadKey(Type_LatLong, 1, 7, 3) == quad_key)
        self.assertEquals("331", quad_key.toInternalString())


def main():
    """
    Bootstrap function for this module
    """
    suite = unittest.TestLoader().loadTestsFromTestCase(TestLatLongSystem)
    unittest.TextTestRunner(verbosity=2).run(suite)
